### ISSUE TITLE
`status` impairs usability in sizeable repos

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#3342]
e.g. git status (0.3sec) vs datalad status (45.8sec):

```shell
fmriprep001$ time git status                                                                                 
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

        new file:   fmriprep/dataset_description.json
        new file:   fmriprep/logs/CITATION.md

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        freesurfer/fsaverage5/surf/.xdebug_tksurfer


real    0m0.253s
user    0m0.184s
sys     0m0.360s
```
NB - those "new file"s are somehow uncommitted after `datalad add` call on all of the files -- yet to possibly figure out reproducible demo
```shell
fmriprep001$ time datalad status                                                                             |
untracked: freesurfer/fsaverage5/surf/.xdebug_tksurfer (file)
    added: fmriprep/dataset_description.json (file)
    added: fmriprep/logs/CITATION.md (file)

real    0m45.858s
user    0m37.752s
sys     0m11.092s
```
so what additional functionality/advantages of `datalad status` demand x100 run time and either it is avoidable?

this dataset is fmriprep output on a dataset with 253 subjects, so there is ~95k files in it (git counted ~200k objects), majority is annexed, no submodules, 

After `git gc --prune=now`, `datalad status` took less -- 34 sec, `git ls-files` takes 0.108 sec, `git ls-tree master -r` also takes whooping 0.1 sec. `datalad save -m '...'` saved those staged files also within just a second or so,  `datalad rev-save` took 45 sec (and did save that `.xdebug_tksurfer` file). 
edit 1: Even when specifying explicitly paths for just a few files to be saved to `rev-save` in a repo with 66k files, requires then a handful of seconds to do a simple commit.

datalad 0.12.0rc2.dev268 (0.12.0rc2-268-ge79d4ea8)  python3

**Comment by yarikoptic** [comment]
as for submodules
<details>
<summary>datalad subdatasets is nicely fast</summary>

```shell
WM-R03$ time datalad subdatasets -r --recursion-limit 1                                                                     
subdataset(ok): BIOS (dataset)
subdataset(ok): BPA2 (dataset)
subdataset(ok): COBY (dataset)
subdataset(ok): DIAMOND (dataset)
subdataset(ok): LAMS (dataset)
subdataset(ok): logs (dataset)
action summary:
  subdataset (ok: 6)

real    0m0.506s
user    0m0.352s
sys     0m0.060s

WM-R03$ time datalad subdatasets -r --recursion-limit 2                                                                     |
subdataset(ok): BIOS (dataset)
subdataset(ok): BIOS/bids (dataset)
subdataset(ok): BIOS/ev3 (dataset)
subdataset(ok): BIOS/fmriprep001 (dataset)
subdataset(ok): BIOS/mriqc (dataset)
subdataset(ok): BIOS/preproc (dataset)
subdataset(ok): BIOS/raw (dataset)
subdataset(ok): BPA2 (dataset)
subdataset(ok): BPA2/bids (dataset)
subdataset(ok): BPA2/ev3 (dataset)
subdataset(ok): BPA2/fmriprep001 (dataset)
subdataset(ok): BPA2/mriqc (dataset)
subdataset(ok): BPA2/preproc (dataset)
subdataset(ok): BPA2/raw (dataset)
subdataset(ok): COBY (dataset)
subdataset(ok): COBY/bids (dataset)
subdataset(ok): COBY/ev3 (dataset)
subdataset(ok): COBY/mriqc (dataset)
subdataset(ok): COBY/preproc (dataset)
subdataset(ok): COBY/raw (dataset)
subdataset(ok): DIAMOND (dataset)
subdataset(ok): DIAMOND/bids (dataset)
subdataset(ok): DIAMOND/ev3 (dataset)
subdataset(ok): DIAMOND/fmriprep001 (dataset)
subdataset(ok): DIAMOND/mriqc (dataset)
subdataset(ok): DIAMOND/mriqc_PRISMA (dataset)
subdataset(ok): DIAMOND/mriqc_TRIO (dataset)
subdataset(ok): DIAMOND/raw (dataset)
subdataset(ok): LAMS (dataset)
subdataset(ok): LAMS/bids (dataset)
subdataset(ok): LAMS/ev3 (dataset)
subdataset(ok): LAMS/fmriprep001 (dataset)
subdataset(ok): LAMS/mriqc (dataset)
subdataset(ok): LAMS/preproc (dataset)
subdataset(ok): LAMS/raw (dataset)
subdataset(ok): logs (dataset)
action summary:
  subdataset (ok: 36)

real    0m0.674s
user    0m0.356s
sys     0m0.148s

WM-R03$ time git status                                                                                                     |
On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)

        modified:   BIOS (modified content)
        modified:   BPA2 (new commits, modified content)
        modified:   COBY (modified content, untracked content)
        modified:   DIAMOND (new commits, modified content, untracked content)
        modified:   LAMS (modified content)

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        .datalad/environments/misc/

no changes added to commit (use "git add" and/or "git commit -a")

real    0m3.701s
user    0m2.308s
sys     0m3.712s
```
</details>

<details>
<summary>git status  is reasonable ~4 sec</summary>

```shell
WM-R03$ time git status                                                                                                     |
On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)

        modified:   BIOS (modified content)
        modified:   BPA2 (new commits, modified content)
        modified:   COBY (modified content, untracked content)
        modified:   DIAMOND (new commits, modified content, untracked content)
        modified:   LAMS (modified content)

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        .datalad/environments/misc/

no changes added to commit (use "git add" and/or "git commit -a")

real    0m3.701s
user    0m2.308s
sys     0m3.712s
```
</details>

and here comes `datalad status` (with a limit to just first few levels)
<details>
<summary>(and this is fast! since on the initial run before I interrupted it after 30 min on initial run before aforementioned commands)</summary>

```shell
WM-R03$ time datalad status -r --recursion-limit 1                                                                          |
untracked: .datalad/environments/misc (directory)                                                                                                            |
 modified: BIOS (dataset)                                                                                                                                    |
 modified: BIOS/ev3 (dataset)                                                                                                                                |
 modified: BIOS/fmriprep001 (dataset)                                                                                                                        |
 modified: BIOS/preproc (dataset)                                                                                                                            |
 modified: BPA2 (dataset)                                                                                                                                    |
 modified: BPA2/ev3 (dataset)                                                                                                                                |
 modified: BPA2/fmriprep001 (dataset)                                                                                                                        |
 modified: COBY (dataset)                                                                                                                                    |
untracked: COBY/fmriprep001 (directory)                                                                                                                      |
untracked: COBY/work (directory)                                                                                                                             |
 modified: COBY/ev3 (dataset)                                                                                                                                |
 modified: COBY/preproc (dataset)                                                                                                                            |
 modified: DIAMOND (dataset)                                                                                                                                 |
^CERROR:                                                                                                                                                     |
Interrupted by user while doing magic: KeyboardInterrupt() [pathlib.py:wrapped:371]
^C^C^C^CException ignored in: <function WeakValueDictionary.__init__.<locals>.remove at 0x7f851bf7ed90>
Traceback (most recent call last):
  File ".../weakref.py", line 109, in remove
    def remove(wr, selfref=ref(self)):
KeyboardInterrupt

real    30m36.724s
user    5m40.820s
sys     1m49.372s
```
</details>

where it proceeded fast until it got stuck on `DIAMOND` on which I interrupted before:
```shell
WM-R03$ time datalad status -r --recursion-limit 1                                                                          |
untracked: .datalad/environments/misc (directory)
 modified: BIOS (dataset)
 modified: BIOS/ev3 (dataset)
 modified: BIOS/fmriprep001 (dataset)
 modified: BIOS/preproc (dataset)
 modified: BPA2 (dataset)
 modified: BPA2/ev3 (dataset)
 modified: BPA2/fmriprep001 (dataset)
 modified: COBY (dataset)
untracked: COBY/fmriprep001 (directory)
untracked: COBY/work (directory)
 modified: COBY/ev3 (dataset)
 modified: COBY/preproc (dataset)
 modified: DIAMOND (dataset)
untracked: DIAMOND/preproc (directory)
 modified: DIAMOND/bids (dataset)
 modified: DIAMOND/ev3 (dataset)
 modified: DIAMOND/fmriprep001 (dataset)
 modified: DIAMOND/mriqc (dataset)
 modified: DIAMOND/mriqc_PRISMA (dataset)
 modified: DIAMOND/mriqc_TRIO (dataset)
 modified: LAMS (dataset)
 modified: LAMS/ev3 (dataset)
 modified: LAMS/fmriprep001 (dataset)
 modified: LAMS/preproc (dataset)

real    8m31.267s
user    5m27.536s
sys     1m15.428s
```

my wild guess is that git is smart to interrupt early when it figures out any of the mentioned in status ` new commits, modified content, untracked content` since it doesn't really need the details for `status`. On a rerun (so everything is 'hot') it took `6m22.557s`

Note: mentioned on that on this hot run it was quite active interleaving (so not taking somehow advantage from running git processes "in parallel" to the datalad process) busy datalad and git, while before it was only at low %s for datalad/git

**Comment by yarikoptic** [comment]
re datasets: `datalad status` takes that long since it traverses the entire hierarchy of datasets even when there is no changes to subdatasets reported by `git status`:
```
WM-R03$ time git status
On branch master
nothing to commit, working tree clean

real    0m3.142s
user    0m2.308s
sys     0m3.580s

WM-R03$ time datalad -l debug status
[DEBUG  ] Command line args 1st pass. Parsed: Namespace() Unparsed: ['status'] 
[DEBUG  ] Discovering plugins 
[DEBUG  ] Building doc for <class 'datalad.core.local.status.Status'> 
[DEBUG  ] Parsing known args among ['/home/amanelis/proj/datalad-master/venvs/dev3/bin/datalad', '-l', 'debug', 'status'] 
[DEBUG  ] Determined class of decorated function: <class 'datalad.core.local.status.Status'> 
[DEBUG  ] Resolved dataset for status reporting: <Dataset path=/data/R03/WM-R03> 
[DEBUG  ] query <AnnexRepo path=/data/R03/WM-R03 (<class 'datalad.support.annexrepo.AnnexRepo'>)>.diffstatus() for paths: None 
[DEBUG  ] <AnnexRepo path=/data/R03/WM-R03 (<class 'datalad.support.annexrepo.AnnexRepo'>)>.get_content_info(...) 
[DEBUG  ] Query repo: ['git', 'ls-files', '--stage', '-z', '-d', '-m', '--exclude-standard', '-o', '--directory', '--no-empty-directory'] 
[DEBUG  ] Done query repo: ['git', 'ls-files', '--stage', '-z', '-d', '-m', '--exclude-standard', '-o', '--directory', '--no-empty-directory'] 
[DEBUG  ] Done <AnnexRepo path=/data/R03/WM-R03 (<class 'datalad.support.annexrepo.AnnexRepo'>)>.get_content_info(...) 
[DEBUG  ] <AnnexRepo path=/data/R03/WM-R03 (<class 'datalad.support.annexrepo.AnnexRepo'>)>.get_content_info(...) 
[DEBUG  ] Query repo: ['git', 'ls-tree', 'HEAD', '-z', '-r', '--full-tree', '-l'] 
[DEBUG  ] Done query repo: ['git', 'ls-tree', 'HEAD', '-z', '-r', '--full-tree', '-l'] 
[DEBUG  ] Initiating a new process for BatchedCommand(cmd=['git', 'cat-file', '--batch'], output_proc=<function>, path='/data/R03/WM-R03') 
[DEBUG  ] Closing stdin of <subprocess.Popen object at 0x7fefd63399e8> and waiting process to finish 
[DEBUG  ] Process <subprocess.Popen object at 0x7fefd63399e8> has finished 
[DEBUG  ] Done <AnnexRepo path=/data/R03/WM-R03 (<class 'datalad.support.annexrepo.AnnexRepo'>)>.get_content_info(...) 
[DEBUG  ] <GitRepo path=/data/R03/WM-R03/BIOS (<class 'datalad.support.gitrepo.GitRepo'>)>.get_content_info(...) 
[DEBUG  ] Query repo: ['git', 'ls-files', '--stage', '-z', '-d', '-m', '--exclude-standard', '-o', '--directory', '--no-empty-directory'] 
[DEBUG  ] Done query repo: ['git', 'ls-files', '--stage', '-z', '-d', '-m', '--exclude-standard', '-o', '--directory', '--no-empty-directory'] 
[DEBUG  ] Done <GitRepo path=/data/R03/WM-R03/BIOS (<class 'datalad.support.gitrepo.GitRepo'>)>.get_content_info(...) 
[DEBUG  ] <GitRepo path=/data/R03/WM-R03/BIOS (<class 'datalad.support.gitrepo.GitRepo'>)>.get_content_info(...) 
[DEBUG  ] Query repo: ['git', 'ls-tree', 'HEAD', '-z', '-r', '--full-tree', '-l'] 
[DEBUG  ] Done query repo: ['git', 'ls-tree', 'HEAD', '-z', '-r', '--full-tree', '-l'] 
[DEBUG  ] Done <GitRepo path=/data/R03/WM-R03/BIOS (<class 'datalad.support.gitrepo.GitRepo'>)>.get_content_info(...) 
[DEBUG  ] <GitRepo path=/data/R03/WM-R03/BIOS/bids (<class 'datalad.support.gitrepo.GitRepo'>)>.get_content_info(...) 
.... and so on until it is satisfied ...
```

**Comment by mih** [comment]
Not sure what it "impairs". You can still use the commands you ran. Also, if you dont care about subdataset state, don't ask for it (`-e`). There already is an issue on caching (#3311), creating one is easy, invalidating it not so much. Looking at the list of subdatasets is insufficient for telling which ones contain uncommitted modifications.

**Comment by yarikoptic** [comment]
I think it might make sense to rely on `git status` to identify modified datasets instead of manual traversal. Could you remind me why it wasn't sufficient?

**Comment by mih** [comment]
Initially, it was direct mode.

**Comment by mih** [comment]
FTR: The code that dives into subdatasets for the sole purpose if figuring out a single "modified" label is here: https://github.com/datalad/datalad/blob/master/datalad/support/gitrepo.py#L3260

Indeed, the recent https://github.com/datalad/datalad/pull/3324 might have had a negative impact on the "full" evaluation mode. Need to look into that.

Otherwise see https://github.com/datalad/datalad/pull/3324#issue-270052649 for a performance comparison of the different modes on a clean dataset hierarchy.

**Comment by mih** [comment]
#3343 Adds early stopping to subdataset state evaluation. This is all we can do, I think, until we can tackle #3311.

**Comment by yarikoptic** [comment]
> > I think it might make sense to rely on git status to identify modified datasets instead of manual traversal. Could you remind me why it wasn't sufficient?

> Initially, it was direct mode.

So, now that there is no direct mode, is there any other reason?
why would we need to reinvent the wheel (ie. not just to use `git status`) and then reinvent caching (#3311) which is already done by git itself for `git status`?  
I am just wondering -- may be there is something we do which git doesn't?

**Comment by mih** [comment]
We are using a variety of queries, starting lower than what `git status` does. Previous attempts to build a system around `git status` failed. One reason is that its internal recursion can only be prevented for the price of not getting any report on submodules. We are reinventing, because we need a solution that fits all demands, not just one aspect.

That being said

**Comment by yarikoptic** [comment]
caching might be a solution, or not... dunno. https://github.com/datalad/datalad/commit/8bf73757960f9c251ab7ad6b44e8eee59d304625 did improve things but  IMHO the issue remains (thus reopening).  Here is another sample
```shell
$ time git status
On branch master
nothing to commit, working tree clean

real    0m0.574s
user    0m0.420s
sys     0m0.760s

$ time datalad status

real    5m32.566s
user    5m23.564s
sys     0m15.064s
```
in a heavy dataset/repository with ~1500 sub-directories on the top level, altogether `ls-tree -r` lists 235270 files/directories. All clean.
Wanted to use `datalad run`, but it seems that the only sensible way would be just with --explicit to avoid overall status check wait (related: #3074 demanding --explicit for rerun)
observation -- `git cat-file --batch` and `datalad status` seems to work nicely in tandem but "serially" consuming ~75% and ~25% CPU accordingly (so no silly sleeps)... probably largely just to "communicate" on 200k files

**Comment by mih** [comment]
It seems it would be sensible to turn of file type detection in run.

**Comment by kyleam** [comment]
> It seems it would be sensible to turn of file type detection in run.

I'm not sure what that would entail, but IMO GitRepo.dirty, which is what `run` is using, doesn't need to use `status()` underneath.  It takes no parameters and returns a boolean, so something along the lines of `<git status --porcelain>.strip()` would suffice.

**Comment by yarikoptic** [comment]
Inspired but a just arrived newbie confusion report about `unsaved modifications present`... Ideally, the assessment of dirtiness and message about unstaged changes should provide some (at least a single file and its status) information about what is changed.  

 E.g. `[unsaved modifications (deleted blah, possibly more) present, save the changes or run in --explicit mode]`

Any command which runs underneath for `is_dirty` check actually receives information about that (even if exits early), so theoretically it should be possible.  Again -- this is an IMHO ideal informative message. Sure thing we could just to tell user to run `datalad status` command to discover what has changed ;-)

**Comment by mih** [comment]
> I'm not sure what that would entail, but IMO GitRepo.dirty, which is what run is using, doesn't need to use status() underneath. It takes no parameters and returns a boolean, so something along the lines of <git status --porcelain>.strip() would suffice.

I think this is a great idea, and I should have done that to begin with (but I was in the mind set of replacing code with uniform pieces).

> Inspired but a just arrived newbie confusion report about unsaved modifications present... Ideally, the assessment of dirtiness and message about unstaged changes should provide some (at least a single file and its status) information about what is changed.

This, unfortunately, directly conflicts with the desire for more speed. I'd prefer to combine @kyleam approach with the second part of your proposal:

> Sure thing we could just to tell user to run datalad status command to discover what has changed

Otherwise we would always pay the price of thorough status evaluation, just for the benefit of being able to deliver an optimal error message fast -- but we should not optimize for fast errors ;-)

**Comment by kyleam** [comment]
>  I'd prefer to combine @kyleam approach with the second part of your proposal:

I've done that in #3460.

> Otherwise we would always pay the price of thorough status evaluation, just for the benefit of being able to deliver an optimal error message fast -- but we should not optimize for fast errors ;-)

I agree, though in my eyes an error message with the status information folded in isn't even an optimal error message.

**Comment by yarikoptic** [comment]
> > Inspired but a just arrived newbie confusion report about unsaved modifications present... Ideally, the assessment of dirtiness and message about unstaged changes should provide some (at least a single file and its status) information about what is changed.
> 
> This, unfortunately, directly conflicts with the desire for more speed. I'd prefer to combine @kyleam approach with the second part of your proposal:

not really, e.g. implementation in #3460 is
```python
return bool(stdout.strip())
```
so `stdout` (output of `git status`) is captured and then discarded just to return a `bool` value.  What I was hinting for is to have a function like `a_dirty_thing` (name could be better ;)) which returns `None` (or another False'y thing, you decide) if there is nothing dirty, and returns first (or a list of items up to some N) dirty items which an then be listed in the message.  So performance wise it would be the same (the same call), but UI would be able to take advantage by providing more informative message.

**Comment by yarikoptic** [comment]
Saving in a sizeable repo (380k files) with 13k files being `git rm`ed already (interrupted `remove` due to #3441 -- didn't have another hour to wait ) takes >14 minutes . process was "busy" for a while talking to `git cat-file --batch` but I stopped eye balling it.

`git status` took 2 sec to report all removed files to be committed.  Didn't check how long it would be just to just `git commit` since either `datalad save` or it  and I decided to time `save`

datalad: 0.12.0rc4-311-gf833c282

**Comment by mih** [comment]
I think with #3701 being merged months ago, this issue can be closed. There is a related one for `diff` #3990
